build_tools/integration_tests/cpu_llm_server_test.py

```diff
@@ -0,0 +1,223 @@
+    start = time.time()
+    while time.time() - start < timeout:
+        try:
+            requests.get(f"{url}/health")
+            return
+        except requests.exceptions.ConnectionError:
+            time.sleep(1)
+    raise TimeoutError(f"Server did not start within {timeout} seconds")
+
+
+@pytest.fixture(scope="module")
+def llm_server(request, model_test_dir, available_port):
+    # Start the server
+    hf_home, tmp_dir = model_test_dir
+    model_file = request.param["model_file"]
+    settings = request.param["settings"]
+    server_process = subprocess.Popen(
+        [
+            "python",
+            "-m",
+            "shortfin_apps.llm.server",
+            f"--tokenizer={hf_home / 'tokenizer.json'}",
+            f"--model_config={tmp_dir / 'edited_config.json'}",
+            f"--vmfb={tmp_dir / 'model.vmfb'}",
+            f"--parameters={hf_home / model_file}",
+            f"--device={settings['device']}",
+        ]
+    )
+    # Wait for server to start
+    wait_for_server(f"http://localhost:{available_port}")
+    yield server_process
+    # Teardown: kill the server
+    server_process.terminate()
+    server_process.wait()
+
+
+def do_generate(prompt, port):
+    headers = {"Content-Type": "application/json"}
+    # Create a GenerateReqInput-like structure
+    data = {
+        "text": prompt,
+        "sampling_params": {"max_tokens": 50, "temperature": 0.7},
+        "rid": uuid.uuid4().hex,
+        "return_logprob": False,
+        "logprob_start_len": -1,
+        "top_logprobs_num": 0,
+        "return_text_in_logprobs": False,
+        "stream": False,
+    }
+    print("Prompt text:")
+    print(data["text"])
+    BASE_URL = f"http://localhost:{port}"
+    response = requests.post(f"{BASE_URL}/generate", headers=headers, json=data)
+    print(f"Generate endpoint status code: {response.status_code}")
+    if response.status_code == 200:
+        print("Generated text:")
+        data = response.text
+        assert data.startswith("data: ")
+        data = data[6:]
+        assert data.endswith("\n\n")
+        data = data[:-2]
+        return data
+    else:
+        response.raise_for_status()
+
+
+@pytest.mark.parametrize(
+    "model_test_dir,llm_server",
+    [
+        (
+            {
+                "repo_id": "SlyEcho/open_llama_3b_v2_gguf",
+                "model_file": "open-llama-3b-v2-f16.gguf",
+                "tokenizer_id": "openlm-research/open_llama_3b_v2",
+                "settings": CPU_SETTINGS,
+                "batch_sizes": [1, 4],
+            },
+            {"model_file": "open-llama-3b-v2-f16.gguf", "settings": CPU_SETTINGS},
+        )
+    ],
+    indirect=True,
+)
+def test_llm_server(llm_server, available_port):
+    # Here you would typically make requests to your server
+    # and assert on the responses
+    assert llm_server.poll() is None
+    output = do_generate("1 2 3 4 5 ", available_port)
+    print(output)
+    assert output.startswith("6 7 8")
```

Comment:
If we were to parameterize this test over the prompt and response too, how much of the model download, import, and compile would be cached between test cases?

(Totally fine to start with the one test case running on every commit, mostly wondering if we can scale this to dozens of test cases running nightly or more frequently)

Reply:
That's a good idea. Currently, the model file and tokenizer file downloaded from hugging face would be cached between runs and re-used.

As of right now, the `mlir` and `vmfb` file would still be regenerated for subsequent runs and saved in another temporary dir